### PR TITLE
add NestMultiply

### PR DESCRIPTION
**Pull request by hnyu**
added a NestCombiner for preprocessing. Will later implement a more general Multiply layer in layers.py

### CODE REVIEW

File: alf/nest/utils.py [hunk 1]
```diff
@@ -111,6 +112,32 @@ def _combine_flat(self, tensors):
...
+        super(NestMultiply, self).__init__(name)
+        if activation is None:
+            activation = lambda x: x
+        self._activation = activation
+
+    def _combine_flat(self, tensors):
+        # Reduce is faster than ``torch.prod(torch.stack(tensors, dim=0), dim=0)``,
+        # and it doesn't require activation to be in-place as what ``torch.prod``
+        # does.
+        ret = functools.reduce(lambda x, y: x * y, tensors,
```
Comment (emailweixu) [inline alf/nest/utils.py]: we have similar math_ops.add_n(). Perhaps we can implement a math_ops.mul_n()

File: alf/utils/math_ops.py [hunk 2]
```diff
@@ -81,49 +82,54 @@ def swish(x):
...
-        the elementwise sum of all the tensors in `inputs`
+        Tensor: the element-wise sum of all the tensors in ``inputs``.
     """
-    assert isinstance(inputs, list)
-    ret = inputs[0]
-    inputs = inputs[1:]
-    for x in inputs:
-        ret = torch.add(ret, x)
-    return ret
+    return functools.reduce(torch.add, inputs, torch.zeros_like(inputs[0]))
```
Comment (emailweixu) [inline alf/utils/math_ops.py]: Are you sure functools.reduce is more efficient?

Comment (emailweixu) [reply alf/utils/math_ops.py]: > Are you sure functools.reduce is more efficient?
Removing initializer might be better:
```
functools.reduce(torch.add, inputs)
```

File: alf/utils/math_ops.py [hunk 3]
```diff
@@ -81,49 +82,54 @@ def swish(x):
...
+    The current solution uses ``functools.reduce``, which is faster than
+    the alternative ``torch.prod(torch.stack(inputs, dim=0), dim=0)``.
+
+    Args:
+        inputs (iterable[Tensor]): an iterable of tensors. It requires that
+            all tensor shapes can be broadcast to the same shape.
+    Returns:
+        Tensor: the element-wise multiplication of all the tensors in ``inputs``.
+    """
+    return functools.reduce(torch.mul, inputs, torch.ones_like(inputs[0]))
```
Comment (emailweixu) [inline alf/utils/math_ops.py]: functools.reduce(torch.mul, inputs)

File: alf/utils/math_ops.py (hunk 2 above)
Comment (hnyu) [reply alf/utils/math_ops.py]: > Are you sure functools.reduce is more efficient?

Most of the time reduce and for-loop should be similar. It's more concise using reduce.

File: alf/utils/math_ops.py (hunk 3 above)
Comment (emailweixu) [reply alf/utils/math_ops.py]: No need to use the initilizer. `functools.reduce(torch.mul, inputs)` is better.

Comment (hnyu) [reply alf/utils/math_ops.py]: already updated

**Review by emailweixu — APPROVED** [review]

**Pull request merged**